mixer/pkg/protobuf/yaml/dynamic/builder.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamic
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+
+	"istio.io/api/policy/v1beta1"
+	"istio.io/istio/mixer/pkg/lang/compiled"
+	"istio.io/istio/mixer/pkg/protobuf/yaml"
+)
+
+type (
+
+	// Compiler compiles expression and returns a value type.
+	Compiler interface {
+		// Compile type Compiler interface {compiles expression and returns a value type.
+		Compile(text string) (compiled.Expression, v1beta1.ValueType, error)
+	}
+
+	// Builder builds encoder based on data
+	Builder struct {
+		msgName     string
+		resolver    yaml.Resolver
+		data        map[interface{}]interface{}
```

Comment:
Nit: If you make data/msgName parameters to build, then you can instantiate this once per yaml.Resolver/compiler.